api: Asset and Task resource types

gioelecerati opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**

Create the new resource types for Assets and Tasks

**Specific updates (required)**

<!--- List out all significant updates your code introduces -->

**How did you test each of these updates (required)**

<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

**Does this pull request close any open issues?**

Fixes [#861](https://github.com/livepeer/livepeer-com/issues/861)
Fixes #880 
Fixes #843 

**Screenshots (optional):**

<!-- Drag some screenshots here, if applicable -->

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

gioelecerati committed:
added schema & table for vodAsset and task

gioelecerati committed:
vod nft property to array

gioelecerati committed:
added audio track parameters & better descriptions

gioelecerati committed:
added userId and deleted fields

gioelecerati committed:
remove id and other fields from required on asset

gioelecerati committed:
remove more required

gioelecerati committed:
extended task object

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: object
+          description: Parameters of the task
+          properties:
+            import:
```
Should we have the params further broken down into `Import`, `Export` and `Transcode` groups, or the `kind` parameter is enough and the params could be on the same level in the object?

victorges replied:
I think the params should be broken down between each `type`, no matter if we have the top-level `params` field or not.

Initially, I had thought about the `import`, `export` and transcode` fields directly on the root of the `task` object, but then they'd probably need a prefix or sufix like `importParams` or `paramsImport`, which looks a bit cumbersome. So I did like having the `params` on the root and then the field with the task type as a key! Looks neat. 

We could do with the params fields directly on the root though, but I do have a strong preference for the params for each separate task contained within their own sub-field (`import`, `export`, etc)

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+                  type: string
+                  description: ID of the asset to transcode
+                  example: 123123123123
+                preset:
```
Should we include transcoding preset adding a Preset resource type in a second step? 

- If yes, should the video and audio transcoding parameters be only in the transcoding preset object and a transcode task takes all the informations from that?
- If no, maybe we should decide which transcoding params are available in a first release

victorges replied:
I think the properties here should be very similar, or maybe 100% equivalent to the ones from `cli-transcoder`. Let's just copy the CLI flags from it and increment this as it grows.

We shouldn't worry about this task type anyway for the MVP. Let's focus on creating the task infra only with the `import` task for now, and we will create the `export` and `transcode` task types right after.

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: string
+          description: Task ID of the task that created the asset
+          example: 1-23-123-123-1123-123
+        relatedTasks:
```
I don't know if it's better to have the `relatedTasks` here in the Asset object to have a view of the task related to it, or if we can display them just with an `assetId` param in the task object, wdyt?

victorges replied:
I think we should have unidirectional links, of children pointing to the parent. So the tasks should have 1 indexed field for the source asset, and if we want to list all related (children) tasks we just query the task table.

Dealing with multiple sources of truth is always a headache, so let's try to keep all of them in a single place.

i.e. IMO we should remove this field

gioelecerati replied:
I agree to remove this field, unidirectional links are more than enough

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+                  type: string
+                  description: Container to use for the transcode
+                  example: mp4
+                audio:
```
Is a transcoding task with audio specification a muted output or a `acodec copy`? Should we specifiy a `muted` param in the audio parameters?

victorges replied:
Let's mimic the `cli-transcoder` for the MVP! If it doesn't support muting audio we might want to add the feature later. But I think it (and Livepeer) does not support audio-transcoding right now, so it will be unlikely that we support it **soon** (might support it sometime tho)

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: string
+          description: Aspect ratio of the asset
+          example: 16:9
+        meta:
```
I think this needs a `MAX_META_SIZE` in the post validation

hjpotter92 replied:
agreed! might need another value to check max size of key-value as well maybe? for eg. total meta size = 1MB, fields inside can be at most 100KB?

victorges replied:
I think there's a JSON schema way to enforce max number of elements in an array. Maybe we can enforce on the keys somehow as well.

iameli commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+              type: number
+              description: height of the profile
+              example: 1080
+            frameRate:
```
I think we call this `fps` elsewhere

hjpotter92 commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: string
+          description:
+            Timestamp (in milliseconds) at which the asset was last updated
+          example: 2022-01-01T02:20:00Z
```
nit: example is not the same as description :sweat_smile:

victorges replied:
All timestamps should be a `number` representing a millisecond-precision Unix timestamp since the epoch, which is also what that `Date.now()` returns by default in node.js. That's our current format in the other objects and we should keep the same pattern here. Mind copying the format/example/etc from the other objects in the schema here?

hjpotter92 commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: string
+          description: Format of the asset
+          example: mp4
+        hash:
```
should it be string or object with key = hash used?

victorges replied:
I also think we should be explicit about what is this hash. For example `md5hash` or `sha256hash`. WDYT
Then we can add other hashes in the future without breaking the contract.

gioelecerati replied:
Something like this looks good?
```
"hash": {
       "hash": "9b560b28b85378a5004117539196ab24e21bbd75b0e9eb1a8bc7c5fd80dc5b57",
        "algorithm": "sha256sum"
}
```  
So we can support different types of hash algorithms wdyt?

hjpotter92 replied:
> 

I think you mean:

```jsonc
"hash": [{
    "hash": "9b560b28b85378a5004117539196ab24e21bbd75b0e9eb1a8bc7c5fd80dc5b57",
    "algorithm": "sha256sum"
},
.
..
]
```
i was going to suggest:

```json
"hash": {
  "sha256sum": "9b560b28b85378a5004117539196ab24e21bbd75b0e9eb1a8bc7c5fd80dc5b57",
  "md5sum": "md5 value here",
  "xxhash": "xxhash value"
}
```

victorges replied:
Like all the ideas!

hjpotter92 commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+        id:
+          type: string
+          description: Task ID
+          example: 1-23-123-123-1123-123
```
should be uuid?

hjpotter92 commented on packages/api/src/schema/schema.yaml:
```diff
@@ -1795,6 +2145,42 @@ paths:
+    post:
+      description: creates a new vod asset
+      responses:
+        200:
```
POST should respond with `201 Created` status ^_^

gioelecerati replied:
Thanks for the catch!

gioelecerati committed:
frameRate to fps

gioelecerati committed:
some changes on examples

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -24,6 +24,8 @@ tags:
     description: Changes user admin status
   - name: webhook
     description: Webhooks endpoints
+  - name: vod-asset
```
I think we're missing the tasks endpoint as well

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+      type: object
+      table: vod_asset
+      required:
+        - name
```
Should also have `additionalProperties: false`. Please add to **every object** on the schema unless there is need for it.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          readOnly: true
+          example: 09F8B46C-61A0-4254-9875-F71F4C605BC7
+          index: true
+        userId:
```
This should also be readonly. Check the other schemas in this files to copy these default fields.

gioelecerati replied:
I saw some of the other schemas have not `readOnly: true` on the `userId` field, only the webhook and the multistream schema have it. Is it intended or it should be added also there?

victorges replied:
Hmm, it should be in all the other `userId` fields yeah! Feel free to add to the existing ones as well, but don't feel obliged. Let's just keep the new ones on the best shape.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+        aspectRatio:
+          type: string
+          description: Aspect ratio of the asset
+          example: 16:9
```
I wonder if we should keep the video-specific fields under a separate subobject here, WDYT? That so this same object could be later used for other file types, like for example images/thumbnails etc.

gioelecerati replied:
I agree, I even think `aspectRatio` is redundant, as we have the `width` and `height` fields in both the track and profile objects?

victorges replied:
Yeah agreed! I think I'm good without that field as well.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: object
+          description: User input metadata associated with the asset
+          example:
+            {
+              "title": "My awesome video",
+              "description": "This is a video of my awesome life",
+              "tags": ["awesome", "life", "video"],
+            }
```
Good vibes example 😆

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: string
+          description: Timestamp (in milliseconds) at which asset was created
+          example: 2022-01-01T02:20:00Z
+        format:
```
This might fit in the `videoParams` or `videoSpec` or something like that.

WDYT of also making this an`enum` with only the formats that we support?

gioelecerati replied:
I think the `format` should stay on the top level of the asset, as this is pretty specific to the container and other fields depend on the container of the asset.

About the enum sounds great, we actually could make separate types for the formats, as they come with different logic / parameters sometimes, wdyt?

victorges replied:
But aren't containers a video-specific thing? So they could be inside the params/spec about `video`-kind of assets, but not necessarily every kind of asset should have a `format`. Even if they do tho, like for example an `image` one might still have a format `jpeg`, the enum values will end up being different anyway, so it's not really the same field, it only has the same name. 

What I am thinking overall is to have the same pattern that we discussed for `Task`s input/output parameters, so that there is both a top-level "type" field which specifies what is the type of the asset, and then there is a specific field with the name of that `type`, with the specific parameters that are only relevant for that specific type. And that can be infinitely nested. 

So for example could have a `type: video` then `video.format: mp4` and then additional `mp4` fields under `video.mp4: {}`, like this:
```
id: 123
name: People Are Awesome
objectStoreId: 345
size: 3476384
type: video
video:
  format: mp4
  width: ...
  height: ...
  bitrate: ...
  mp4:
    anyCustomMp4Stuff: ...
```

That way, on the root of the `asset` object, we only hold directly the fields that would be valid for almost any kind of "file" in general. Then we specify what type of file it is, then have a dedicated sub-object for that type, etc.

On our MVP we will only have `video` files. But soon we might already have `hls` as a separate `type` altogether (or just as a separate `format`? Maybe not since it's not even a single file). And later we might want to have other things like `image` perhaps, for thumbnails or even GIFs or something. Maybe eventually `audio` as well for audio-only formats? Or would those be stored as "video"?

WDYT?

gioelecerati replied:
Sounds good, we could put all these fields and the ones you cited in the other comments under a video field. It is for sure clearer, if we plan on using the same Asset for other type of assets (images, NFT, audios..)  
At this point I would rename from `vod-asset` to `asset`, as vod is video specific, wdyt?
I also wonder if in a second step we should separate all the metadata in a separate place in the schema

victorges replied:
Just asset instead of vod-asset makes sense to me as well! And we can separate the metadata, but I'm ok with keeping it on the root now as, after all, this whole object is metadata anyway. So we got the common file metadata on the root and the specialized metadata in the category-specific fields.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+          type: string
+          description: Hash of the asset
+          example: sha256:hu31498hf14398
+        tracks:
```
Another one for `videoSpec`

gioelecerati replied:
Are you referring to `tracks`? This is related not only to the videoSpec, as tracks is an array that can contain also audio tracks (e.g. MP4 container with 1 video track and multiple audio tracks - audio languages)

victorges replied:
I tried to explain better what I mean by that `videoSpec` here: https://github.com/livepeer/livepeer-com/pull/865#discussion_r790007068

I know that `tracks` can also be audio, but at least in this context I understood that they are tracks in a video-file / container. So they could be under the specific "video" segment perhaps (or something else that we decide to call it).

What would happen if/when we supported audio-files for example? Would they re-use all this same fields, with the only difference that it only has audio tracks? So what could be an asset type that could be either a "video"-file or an audio-file? We still have the case for images tho, which definitely don't have `tracks` (right? 😳), and so this field wouldn't make sense for them.

Idk, check my other comment first that I linked here, maybe the idea I'm suggestion does not make that much sense if the things are not that well separated in the end.

gioelecerati replied:
I put also `tracks` under the video field (as well as `profiles`). I guess we can later decide if all the specs should be separate and linked to asset, so the `asset` object becomes totally agnostic to what kind of asset it is

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+        tracks:
+          type: array
+          description:
+            List of tracks associated with the asset when the format contemplate
```
**nit:** typo

```suggestion
            List of tracks associated with the asset when the format contemplates
```

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+            List of tracks associated with the asset when the format contemplate
+            them (e.g. mp4)
+          items:
+            type: object
```
Need `additionalProperties: false` as well. Could have a `required: [kind]` as well, or including any other fields that are always present.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+              codec:
+                type: string
+                description: Codec of the track
+                example: aac
```
Is this only for audio as well?

Since the are so different in the specs, should we have separate `audioTracks` and `videoTracks` perhaps? Or would that be counter intuitive considering how other tools and platforms represent this?

gioelecerati replied:
Yeah as I mentioned in another comment, a track can be either a audio or a video track. For example, this command:
```
ffprobe -v quiet -print_format json -show_format -show_streams example.mp4
```
would output a similar json, with one or more separate streams (`tracks`), either audio or video. In this case all the params are even the same for both the tracks.  
In an example case, `codec` could be `aac` or `h264` depending on which track it is

victorges replied:
Hmm I see. Yeah so let's follow the standard from `ffmpeg` then. I guess it could be useful for "newcomers" to have tracks explicitly separated about what are the audio and video tracks, but maybe it would feel alien for more experienced video-developers 😆

victorges reviewed (commented):
Some preliminary comments just in case GitHub app loses the local state

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+                type: string
+                description: Codec of the track
+                example: aac
+        profiles:
```
Now this should be a field under an `hlsSpec` instead!

Possibly `hlsSpec` lives inside `videoSpec` as well, in case it still uses many of the video fields as well.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+                description: Codec of the track
+                example: aac
+        profiles:
+          type: object
```
Gotta disable additional properties

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+        playbackUrl:
+          type: string
+          description: Playback url of the playlist manifest
+          example: https://cdn.livepeer.com/ABC123/playlist.m3u8
```
We should follow the pattern from streams and only store the playbackId as well. The url should be built and could change over time depending on the infra. From us getting a new domain to even changing the url structure somehow.

gioelecerati committed:
added additionalProperties - task endpoints - hash as object

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +813,354 @@ components:
+              type: number
+              description: Bitrate of the profile in bits per second
+              example: 1000000
+        playbackUrl:
```
This should be under `hlsSpec` as well (or just `hls` perhaps? Maybe we can drop the `Spec` suffix on these sub-objects)

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,376 @@ components:
+              example: ABC123
+            duration:
+              type: number
+              description: Duration of the profile in seconds
```
Also `readOnly`. Actually I think the whole asset spec is readonly right? Users won't be able to modify an asset via a PUT or something like that, only through specific tasks which then CRUD the asset 🤔

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,376 @@ components:
+              description: ID of the profile
+              example: ABC123
+            duration:
+              type: number
```
Instead of a single example, I think this should have/be an `enum` instead and we list all the possible values here

gioelecerati committed:
added bunch of readOnly

gioelecerati committed:
fixed examples

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+            ID of the source asset (root) - If missing, this is a root asset
+          example: 09F8B46C-61A0-4254-9875-F71F4C605BC7
+        nft:
+          type: array
```
I think an nft should be just another immutable asset pointing to the file on a decentralized storage, with whatever extra nft parameters saved on an equivalent "nft spec" field (which could be `nftSpec` or just `nft` as I'm realizing now).

If there are multiple NFTs minted from the same asset, there would be multiple separate assets created, each with their own NFT spec (but possibly the same storage?)

victorges replied:
Actually, for now let's just remove this field. We will not be handling NFTs directly on the first version (maybe soon, but not on the import/export MMMVP) and we shouldn't have fields in the schema that we don't have the code for. 

When we do implement it, everything will also be clearer and we should know better how to model it on top of the existing asset. (was useful having it here to think about/raise discussion tho)

gioelecerati replied:
I agree that NFTs should be another asset, let's remove it and implement it later

victorges reviewed (commented):
Finishing the asset schema now

gioelecerati committed:
fix typo

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+        kind:
+          type: string
+          description: Type of the task
+          example: import
```
I think `kind` already has a legacy/tech debt meaning in some of the other resources, which like just stores what is the API resource of the object. I think it's a reminiscent from when we did not use PostgreSQL with different tables for each API resource, but maybe a single Firestore "table" if all the objects from the API. So for example all `stream` objects have a `kind: "stream"`, all `user` objects have a `kind: "user"` etc.

Soo, to avoid confusion, I think we might want to use `type` here instead of a`kind`.

Also, we should have an `enum` here for the exact types of task that we support. As we add more task types, we append to the `enum` here as well.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+        status:
+          type: string
+          description: Status of the task
+          example: ready
```
This should/could also be an `enum` instead of only an `example`

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+        updatedAt:
+          type: string
+          description: Timestamp (in milliseconds) at which task was updated
+          example: 1587667174725
```
These should be `number`s as well

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+            Timestamp (in milliseconds) at which the asset was last updated
+          example: 1587667174725
+        createdAt:
+          type: string
```
Forgot to update the type to `number`.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+                  type: string
+                  description: URL of the asset to import
+                  example: https://cdn.livepeer.com/ABC123/filename.mp4
+            export:
```
Same as NFT, let's not create the schemas for these until we actually start implementing them. Otherwise they might end up with extraneous data or fields that we don't use or even notice.

I'd say we should start with only `import`, then right after we will implement `export` and maybe later `transcode`.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+              additionalProperties: false
+              description: Parameters for the export task
+              properties:
+                kind:
```
WDYT of `protocol` or `service` instead of `kind`? I think "kind" would be too generic here

gioelecerati replied:
I removed the export task for now to define it later when we implement it!

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+                kind:
+                  type: string
+                  description: Type of the export
+                  example: arweave|ipfs|s3
```
Consider an `enum` as well, I guess we'll have to implement specific support for any additional protocol anyway, so could have the supported ones enumerated here as well.

victorges replied:
Also, let's keep the enum to only `s3` for now if that's the only one we are going to support.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+                      type: number
+                      description: Bit depth of the audio
+                      example: 16
+                video:
```
I **think** `cli-transcoder` only supports some preset profiles for now.  We don't need to check that now anyway, let's create this with only the import task.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+              type: object
+              additionalProperties: false
+              description: Parameters for the export task
+              properties:
```
Might need an `assetId` / `sourceAssetId` on this one as well?

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -1795,6 +2174,42 @@ paths:
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/error"
```
I think we won't have a raw "create" like this. You can only create an asset from an import task or by uploading a new file. Uploading a file might require a little protocol like calling the API, receiving a pre-signed S3 URL to upload the file, then creating the `asset` with the link/pre-signed for that S3 file. So we might want to have separate/explicit API endpoints like `POST /assets/upload` perhaps which returns the S3 link, then another call which will basically import that file to the regular object store? The "import" would be only so that we run the `ffprobe` on the file and fill the asset spec, but could avoid the actual download&upload from S3 but only renaming the file which would already be in our own bucket. WDYT?

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,381 @@ components:
+      type: object
+      table: task
+      additionalProperties: false
+      properties:
```
I think some fields are missing:
 - An indexed `sourceAssetId`, so that we can always search for tasks related to a given task. It would work differently from the `sourceAssetId` of the assets themselves tho, pointing to the direct parent and not to the tree root, so we might want to call it something else. I think `parentAssetId` could make sense then. It can also be `null` in case of "import" tasks for example. In the future it might even be a list of assets, in case of stitching 2 videos together for example, but for now I think we're good with a single parent ID, which will be simpler (it's kinda complex to get indexes on JSON arrays as well (tho definitely possible))
 - The output of the task. I think on our design we had thought of a `status` **object** instead of just a string, and inside the status object we would have both the current "phase" (k8s-inspired) which is that `Pending`/`Running` etc, a timestamp of the last change, and then the `output` of the task after it's down (which should follow the same pattern as `params` IMO, with the separate sub-fields for each task type). So I think the current `status` and `updatedAt` should be moved to an actual `status` object (`status`=>`phase`) and then we add the `output` there as well. WDYT?

victorges reviewed (commented):
now some ideas on the task schema!

gioelecerati committed:
changes after review

gioelecerati committed:
changed hash to array, fixed timestamp type

gioelecerati committed:
added parentAssetId and output fields to task object

gioelecerati committed:
enums

gioelecerati committed:
rename vod-asset to asset

gioelecerati committed:
draft task ctrl

gioelecerati committed:
removed task test - added owner to task

iameli commented on packages/api/src/controllers/task.js:
```diff
@@ -0,0 +1,180 @@
+});
+
+app.get("/:id", authMiddleware({}), async (req, res) => {
+  const os = await req.store.get(`task/${req.params.id}`);
```
Should have `db.task` in all cases

iameli commented on packages/api/src/controllers/task.js:
```diff
@@ -0,0 +1,180 @@
+  } catch (e) {
+    console.error(e);
+    throw e;
+  }
```
Shouldn't need the try/catch, will just get picked up by our middleware

```suggestion
  await db.task.create(doc);
```

iameli commented on packages/api/src/controllers/task.js:
```diff
@@ -0,0 +1,180 @@
+  res.end();
+});
+
+app.put("/:id", authMiddleware({}), validatePost("task"), async (req, res) => {
```
Maybe we just need an admin-only `PATCH` endpoint for the worker to update

victorges replied:
Like this idea, at least for a start. Same for the assets, which at first can/may be editable only by the task runner itself (apart from very surface metadata like name/title).

gioelecerati committed:
req.store to db - removed try

gioelecerati committed:
js to ts - fixed deleted / fixed db.task.get

gioelecerati committed:
update to patch - only admin

gioelecerati committed:
added asset controller

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,289 @@ components:
+        deleted:
+          type: boolean
+          description: Set to true when the asset is deleted
+        objectStoreId:
```
So I found that this field is considered "private" on the streams, which is likely because most of the time it will be referencing an object that cannot be retrieved by the user (an object store in Ivan's account). I believe we should follow the same pattern here, and make this field a `writeOnly` property, which will never be returned by the API **unless** an admin is making the request (so that `task-runner` can access it).

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,289 @@ components:
+              additionalProperties: false
+              description: Output of the import task
+              properties:
+                status:
```
The `status` field should be shared by all task types instead. I think it's more of a metadata than an output instead as well, so I think it can be moved to the root. 

Following the Kubernetes API, the status could also be an object that has a `phase` of `Pending/Running/Failed/Completed` instead, and we can move the `updatedAt` to inside the status as well WDYT?

```
status:
  phase: Pending
  updatedAt: 12437328123
  // later we can add more fields if appropriate
```

gioelecerati replied:
Yeah, I agree, we could move it to the root, I was thinking that different kinds of task in the future could have new types of status phases, but I guess we can add them in the phase enum and support all cases

gioelecerati committed:
added vodObjectStore id param

gioelecerati committed:
added playbackId - objectStoreId writeOnly

gioelecerati committed:
added aws-sdk client and presigner

gioelecerati committed:
added presigned url helper - request-upload and upload proxy asset endpoints

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,294 @@ components:
+                required:
+                  - kind
+                properties:
+                  kind:
```
For consistency with ffmpeg and the rest of the schema we're avoiding `kind`, maybe this should be `type` instead

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,294 @@ components:
+                    type: string
+                    description: Codec of the track
+                    example: aac
+            hlsSpec:
```
Let's keep this off/commented until we actually implement it as well

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,294 @@ components:
+              description: HLS spec for the asset
+              properties:
+                profiles:
+                  type: object
```
I think you meant array of objects

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,294 @@ components:
+          description: Object store ID where the asset is stored
+          writeOnly: true
+          example: 09F8B46C-61A0-4254-9875-F71F4C605BC7
+        name:
```
Hmm, I think we will want to get the filename somehow on the import asset flow in the end to populate this. Or should it be populated on the request to start the import instead? Maybe both, I'll do it on the import task as well since it's easy but the user can provide one on the request to start the import, which if provided should be used instead of the one that the import task gets.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,294 @@ components:
+            type: object
+            additionalProperties: false
+            properties:
+              hash:
```
Reminder for myself that I need a third processing of the file contents for calculating this hash. Or get it from the response from SE, but I'm not sure if it's available (currently not exposed on the OS drivers lib at least)

gioelecerati committed:
review changes

gioelecerati committed:
removed check

gioelecerati committed:
added update status endpoint

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,310 @@ components:
+      additionalProperties: false
+      required:
+        - name
+      properties:
```
I think we are missing a `size` property on the root of the asset object as well.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,310 @@ components:
+                    description: Frame rate of the track - only for video tracks
+                    example: 30
+                  id:
+                    type: string
```
What is this track ID? ffprobe apparently doesn't return anything like that.

gioelecerati replied:
Yeah we don't need it, there's not reference to it elsewhere, removed!

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -811,6 +815,310 @@ components:
+                    type: number
+                    description: Height of the track - only for video tracks
+                    example: 1080
+                  fps:
```
I think we should also have the `pixelFormat` here since it's one of the things that can break the transcoding flow (`pixfmt` from ffmpeg)

gioelecerati replied:
Good point!

gioelecerati committed:
added size, pixelFormat, removed trackid

gioelecerati committed:
added proxy callback and task creation

gioelecerati committed:
added tasktrigger and taskresult

gioelecerati committed:
added exchange and queue

gioelecerati committed:
added http-proxy types

gioelecerati committed:
changed queue binding

gioelecerati committed:
added import - changed publish

gioelecerati committed:
taskId in the event key

gioelecerati committed:
added uploadedObjectKey - objectStoreId on asset gen

gioelecerati committed:
edits

victorges committed:
api: Sync asset schema with task runner

victorges commented on packages/api/src/controllers/asset.ts:
```diff
@@ -0,0 +1,351 @@
+
+function validateAssetPayload(id, userId, createdAt, payload) {
+  try {
+    if (payload.meta && JSON.stringify(payload.meta).length > META_MAX_SIZE) {
```
nice :)

victorges committed:
api: Sync a few bits and pieces on assets controller

Mainly:
 - Fixed a couple typescript issues which then showed a few bugs
 - Fix the schema of the task events, now type-checked on callsite
 - Add a couple missing fields in the creation of assets, making sure we
   always have a playbackId in a format consistent with streams'
 - Change a few things on the schema, mainly adding the right status
   enums and making all values lowercase to match the pattern in our
   other "status/type-like" fields (like stream.status itself)

victorges committed:
api/queue: Fix task exchange and queue names

victorges committed:
api: More syncs for tasks to start running!

Mainly:
 - Add object store ID to every asset
 - Update task status API

victorges committed:
api: Remove a couple forgotten logs

gioelecerati committed:
added TaskScheduler

gioelecerati committed:
log

gioelecerati committed:
fixed presignedUrl parsing on upload

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -0,0 +1,95 @@
+            );
+          }
+
+          db.asset.update(task.parentAssetId, {
```
The task status also needs to be updated with this output.

gioelecerati committed:
added task update phase status

victorges commented on packages/api/src/controllers/asset.ts:
```diff
@@ -0,0 +1,431 @@
+  const lpSignedUrl = `https://${req.frontendDomain}/api/asset/upload/${b64SignedUrl}`;
+
+  await db.asset.create({
+    id,
```
Seems to be missing a couple of fields here. Either use the `validateAssetPayload` which I think should become a function to "create new asset payload with all the required fields". Or you can repeat them here, but I think it's worth keeping all those default stuff in a single place, since there will be a lot of new asset creation flows to come.

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -0,0 +1,102 @@
+          }
+
+          db.asset.update(task.parentAssetId, {
+            hash: assetSpec.hash,
```
need `await` here and above

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -0,0 +1,102 @@
+              `task event process error: assetSpec not found in TaskResult for task ${event.task.id}`
+            );
+          }
+
```
could add a TODO here to do these in a transaction

gioelecerati committed:
output metadata - added indexes

gioelecerati committed:
fix await on result

victorges committed:
api: Fix object-store tests

gioelecerati committed:
lockfile

gioelecerati committed:
Merge branch 'vod-api' of github.com:livepeer/livepeer-com into vod-api

gioelecerati committed:
prettier

victorges reviewed (approved):
Ship it!
